### PR TITLE
4556 - improve formatparameter

### PR DESCRIPTION
**Pull request by luu-alex**
#4556 input does not mutate in formatparameter anymore

### CODE REVIEW

**Review by avkos — APPROVED** [review]

File: packages/web3-eth-abi/src/utils.ts [hunk 1]
```diff
@@ -123,12 +123,18 @@ export const mapTypes = (
...
+export const formatOddHexstrings = (param: string): string =>
+	param.length % 2 === 1 ? `0x0${param.substring(2)}` : param;
+
 /**
  * Handle some formatting of params for backwards compatibility with Ethers V4
  */
 export const formatParam = (type: string, _param: unknown): unknown => {
 	// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
-	let param = _param;
+	const param = _param;
```
Comment (jdevcs) [inline packages/web3-eth-abi/src/utils.ts]: More safe approach is to clone `_param` obj, as `param` is also returned in this function.

File: packages/web3-eth-abi/src/utils.ts [hunk 2]
```diff
@@ -169,18 +173,16 @@ export const formatParam = (type: string, _param: unknown): unknown => {
...
+					? rightPad(param as string, size * 2)
+					: hexParam;
+			return formatOddHexstrings(paddedParam as string);
 		}
 
-		// format odd-length bytes to even-length
-		if ((param as string).length % 2 === 1) {
-			param = `0x0${(param as string).substring(2)}`;
-		}
+		return formatOddHexstrings(hexParam as string);
```
Comment (jdevcs) [inline packages/web3-eth-abi/src/utils.ts:193]: I think good to add 'input does not mutate' unit test.

Comment (nazarhussain) [reply packages/web3-eth-abi/src/utils.ts:193]: If input mutate semantically, the TS and ESLint will complaint.

File: packages/web3-eth-abi/src/utils.ts [hunk 3]
```diff
@@ -123,12 +123,18 @@ export const mapTypes = (
 	return mappedTypes;
 };
 
+/**
+ * format odd-length bytes to even-length
+ */
+export const formatOddHexstrings = (param: string): string =>
+	param.length % 2 === 1 ? `0x0${param.substring(2)}` : param;
+
```
Comment (nazarhussain) [inline packages/web3-eth-abi/src/utils.ts]: As it's a utility function exported from package, so would be nice to validate the input it's really a hex string and odd lenght.

**Review by nazarhussain — CHANGES_REQUESTED** [review]
Apart from @jdevcs feedback LGTM.

File: packages/web3-eth-abi/src/utils.ts [hunk 4]
```diff
@@ -123,12 +123,26 @@ export const mapTypes = (
...
 /**
  * Handle some formatting of params for backwards compatibility with Ethers V4
  */
 export const formatParam = (type: string, _param: unknown): unknown => {
 	// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
-	let param = _param;
+
+	// clone if _param is an object
+	const param = typeof _param === 'object' && !Array.isArray(_param) ? { ..._param } : _param;
```
Comment (jdevcs) [inline packages/web3-eth-abi/src/utils.ts:145]: @luu-alex  This PR LGTM, one observation: Even its array we should clone it I think. or is there any reason for not doing that?

**Review by nazarhussain — APPROVED** [review]

File: packages/web3-eth-abi/src/utils.ts (hunk 4 above)
Comment (luu-alex) [reply packages/web3-eth-abi/src/utils.ts:145]: nevermind, i updated to cloning if its an array

Comment (luu-alex) [reply packages/web3-eth-abi/src/utils.ts:145]: https://github.com/ChainSafe/web3.js/blob/e8ce18b293ebb1f848d3d13ac4ed1ac5a1a108e8/packages/web3-eth-abi/src/utils.ts#L158 updated

**Review by jdevcs — APPROVED** [review]

**Pull request merged**